node/runtime/pangoro/src/pallets/bridge_dispatch.rs

```diff
@@ -1,19 +1,187 @@
+	if old_remaining < amount_remaining {
+		// Add 1, if the substrate balance part touched
+		amount_sub = amount_sub.saturating_add(U256::from(1));
+	}
+
+	// Calculate the new substrate balance part
+	let new_evm_balance = old_evm_balance.saturating_sub(amount);
+	let (new_sub, _new_remaining) = new_evm_balance.div_mod(U256::from(POW_9));
+
+	// Ensure the account underlying substrate account has no liquidity restrictions.
+	ensure!(
+		Ring::ensure_can_withdraw(
+			who,
+			amount_sub.low_u128().unique_saturated_into(),
+			reasons,
+			new_sub.low_u128().unique_saturated_into(),
+		)
+		.is_ok(),
+		TransactionValidityError::Invalid(InvalidTransaction::Payment)
+	);
+
+	Ok(())
+}
+
+pub struct CallValidator;
+impl CallValidate<bp_pangoro::AccountId, Origin, Call> for CallValidator {
+	fn check_receiving_before_dispatch(
+		relayer_account: &bp_pangoro::AccountId,
+		call: &Call,
+	) -> Result<(), &'static str> {
+		match call {
+			Call::Ethereum(darwinia_ethereum::Call::message_transact { transaction: tx }) =>
+				match tx {
+					Transaction::Legacy(t) => {
+						// Use fixed gas price now.
+						let gas_price =
+							<Runtime as darwinia_evm::Config>::FeeCalculator::min_gas_price();
+						let fee = t.gas_limit.saturating_mul(gas_price);
+
+						// Ensure the relayer's account has enough balance to withdraw. If not,
+						// reject the call.
+						ensure!(
+							evm_ensure_can_withdraw(
+								relayer_account,
+								cmp::min(
+									fee,
+									decimal_convert(MaxUsableBalanceFromRelayer::get(), None)
+								),
+								WithdrawReasons::TRANSFER
+							)
+							.is_ok(),
+							TransactionValidityError::Invalid(InvalidTransaction::Payment)
+						);
+						Ok(())
+					},
+					_ => Ok(()),
+				},
+			_ => Ok(()),
+		}
+	}
+
+	fn call_validate(
+		relayer_account: &bp_pangoro::AccountId,
+		origin: &Origin,
+		call: &Call,
+	) -> Result<(), TransactionValidityError> {
+		match call {
+			// Note: Only supprt Ethereum::message_transact(LegacyTransaction)
+			Call::Ethereum(darwinia_ethereum::Call::message_transact { transaction: tx }) => {
+				match origin.caller() {
+					OriginCaller::Ethereum(RawOrigin::EthereumTransaction(id)) => match tx {
+						Transaction::Legacy(t) => {
+							// Only non-payable call supported.
+							if t.value != U256::zero() {
+								return Err(TransactionValidityError::Invalid(
+									InvalidTransaction::Payment,
+								));
+							}
+
+							// Use fixed gas price now.
+							let gas_price =
+								<Runtime as darwinia_evm::Config>::FeeCalculator::min_gas_price();
+							let fee = t.gas_limit.saturating_mul(gas_price);
+
+							// Ensure the relayer's account has enough balance to withdraw.
+							if evm_ensure_can_withdraw(
+								relayer_account,
+								cmp::min(
+									fee,
+									decimal_convert(MaxUsableBalanceFromRelayer::get(), None),
+								),
+								WithdrawReasons::TRANSFER,
+							)
```

Comment:
ensure fee <= decimal_convert(MaxUsableBalanceFromRelayer::get(), None)

Reply:
The check here seems duplicated with the validation in the `check_receiving_before_dispatch`?

Reply:
Check my latest commit [57344c1](https://github.com/darwinia-network/darwinia-common/pull/1267/commits/57344c1d2c4b1aa201d6a35f0d663919306f6cd5)